packages/flutter/lib/src/widgets/autocomplete.dart

```diff
@@ -0,0 +1,579 @@
+    required this.fieldViewBuilder,
+    required this.optionsViewBuilder,
+    required this.optionsBuilder,
+    this.displayStringForOption = _defaultStringForOption,
+    this.onSelected,
+  }) : assert(fieldViewBuilder != null),
+       assert(optionsViewBuilder != null);
+
+  /// Builds the field whose input is used to get the options.
+  ///
+  /// Pass the provided [TextEditingController] to the field built here so that
+  /// AutocompleteCore can listen for changes.
+  final AutocompleteFieldViewBuilder fieldViewBuilder;
+
+  /// Builds the selectable options widgets from a list of options objects.
+  ///
+  /// The options are displayed floating below the field using a
+  /// [CompositedTransformFollower] inside of an [Overlay], not at the same
+  /// place in the widget tree as AutocompleteCore.
+  final AutocompleteOptionsViewBuilder<T> optionsViewBuilder;
+
+  /// Returns the string to display in the field when the option is selected.
+  ///
+  /// This is useful when using a custom T type and the string to display is
+  /// different than the string to search by.
+  ///
+  /// If not provided, will use `option.toString()`.
+  final AutocompleteOptionToString<T> displayStringForOption;
+
+  /// Called when an option is selected by the user.
+  ///
+  /// Any [TextEditingController] listeners will not be called when the user
+  /// selects an option, even though the field will update with the selected
+  /// value, so use this to be informed of selection.
+  final AutocompleteOnSelected<T>? onSelected;
+
+  /// A function that returns the current selectable options objects given the
+  /// current TextEditingValue.
+  final AutocompleteOptionsBuilder<T> optionsBuilder;
+
+  // The default way to convert an option to a string.
+  static String _defaultStringForOption(dynamic option) {
+    return option.toString();
+  }
+
+  @override
+  _AutocompleteCoreState<T> createState() => _AutocompleteCoreState<T>();
+}
+
+class _AutocompleteCoreState<T extends Object> extends State<AutocompleteCore<T>> {
+  final GlobalKey _fieldKey = GlobalKey();
+  final LayerLink _optionsLayerLink = LayerLink();
+  final TextEditingController _textEditingController = TextEditingController();
+  Iterable<T> _options = Iterable<T>.empty();
+  T? _selection;
+
+  // The OverlayEntry containing the options.
+  OverlayEntry? _floatingOptions;
+
+  // True iff the state indicates that the options should be visible.
+  bool get _shouldShowOptions {
+    final TextSelection selection = _textEditingController.selection;
+    final bool fieldIsFocused = selection.baseOffset >= 0
+        && selection.extentOffset >= 0;
+    final bool hasOptions = _options.isNotEmpty;
+    return fieldIsFocused && _selection == null && hasOptions;
+  }
+
+  // Called when _textEditingController changes.
+  void _onChangedField() {
+    final Iterable<T> options = widget.optionsBuilder(
+      _textEditingController.value,
+    );
+    assert(options != null);
```

Comment:
Could produce a more informative error message here; explain that the optionsBuilder callback returned null ... but is there a runtime exception in this case? The type of the callback is ` AutocompleteOptionToString<T>` not  `AutocompleteOptionToString<T?>`. 

Similar feedback (question really) about the other callbacks that return `T` but check the return value for null?

Reply:
You're right, these assertions for null are unnecessary.  I'll remove them.  Thanks for catching that!